lib/rspec/core/formatters/snippet_extractor.rb

```diff
@@ -1,114 +1,133 @@
         end
 
-        # rubocop:enable Style/ClassVars
-
-        # @api private
-        #
-        # Extract lines of code corresponding to  a backtrace.
-        #
-        # @param backtrace [String] the backtrace from a test failure
-        # @return [String] highlighted code snippet indicating where the test
-        #   failure occured
-        #
-        # @see #post_process
-        def snippet(backtrace)
-          raw_code, line = snippet_for(backtrace[0])
-          highlighted = @@converter.convert(raw_code)
-          post_process(highlighted, line)
-        end
-        # rubocop:enable Style/ClassVars
-
-        # @api private
-        #
-        # Create a snippet from a line of code.
-        #
-        # @param error_line [String] file name with line number (i.e.
-        #   'foo_spec.rb:12')
-        # @return [String] lines around the target line within the file
-        #
-        # @see #lines_around
-        def snippet_for(error_line)
-          if error_line =~ /(.*):(\d+)/
-            file = Regexp.last_match[1]
-            line = Regexp.last_match[2].to_i
-            [lines_around(file, line), line]
-          else
-            ["# Couldn't get snippet for #{error_line}", 1]
+        if RSpec::Support::RubyFeatures.ripper_supported?
+          NoExpressionAtLineError = Class.new(StandardError)
+
+          PAREN_TOKEN_TYPE_PAIRS = {
+            :on_lbracket    => :on_rbracket,
+            :on_lparen      => :on_rparen,
+            :on_lbrace      => :on_rbrace,
+            :on_heredoc_beg => :on_heredoc_end
+          }
+
+          attr_reader :source, :beginning_line_number, :max_line_count
+
+          def self.extract_expression_lines_at(file_path, beginning_line_number, max_line_count=nil)
+            if max_line_count == 1
+              [extract_line_at(file_path, beginning_line_number)]
+            else
+              source = source_from_file(file_path)
+              new(source, beginning_line_number, max_line_count).expression_lines
+            end
           end
-        end
 
-        # @api private
-        #
-        # Extract lines of code centered around a particular line within a
-        # source file.
-        #
-        # @param file [String] filename
-        # @param line [Fixnum] line number
-        # @return [String] lines around the target line within the file (2 above
-        #   and 1 below).
-        def lines_around(file, line)
-          if File.file?(file)
-            lines = File.read(file).split("\n")
-            min = [0, line - 3].max
-            max = [line + 1, lines.length - 1].min
-            selected_lines = []
-            selected_lines.join("\n")
-            lines[min..max].join("\n")
-          else
-            "# Couldn't get snippet for #{file}"
+          def self.source_from_file(path)
+            raise_error_if_file_does_not_exist!(path)
+            @source_cache ||= {}
+            @source_cache[path] ||= Source.from_file(path)
           end
```

Comment:
I'd prefer we not attach global mutable state to the class like this.  Issues I have with it:

* It can make testing difficult.  You have to consciously think about clearing this source cache between tests.
* I think users expect that all RSpec global state will be cleared when `RSpec.reset` is called, but this state here will not be cleared as currently implemented.  Concretely, this can cause problems when the user uses an alternate runner that keeps the RSpec process around and then triggers it to load and run spec files via a signal or some other mechanism.  In such a situation, the source files could change between runs but the cache here would persist and could lead to the wrong snippets being displayed.
* I think it makes the logic harder to reason about.

Instead, what do you think of this alternate design?

* Remove the global state from the snippet extractor.
* Add a `Source::Cache` class (presumably defined in `source.rb`) that exposes a `source_from_file` instance method.
* Expose a `Source::Cache` instance off of some other RSpec object that gets automatically cleared when RSpec is reset.  I'm not sure the best place but the reporter (e.g. `RSpec.configuration.reporter.source_cache`) and `RSpec.world.source_cache` are two possibilities.  Neither of those is obviously perfect so there may be a better place I'm not thinking of.